examples/gaussian_process/svgp.py

```diff
@@ -0,0 +1,216 @@
+
+
+def rbf(x, y, inv_scale):
+    x = tf.expand_dims(x, 1)
+    y = tf.expand_dims(y, 0)
+    inv_scale = tf.reshape(inv_scale, [1, 1, -1])
+    return tf.exp(- tf.reduce_sum(tf.square(x - y) * inv_scale, axis=-1) / 2)
+
+
+@zs.reuse('model')
+def build_model(observed, Z_pos, X_ph, hps, n_particles, full_cov=False):
+    '''
+    Build the SVGP model.
+    Note that for inference, we only need the diagonal part of Cov[Y], as ELBO
+    equals sum over individual observations.
+    For visualization etc we may want a full covariance. Thus the argument
+    `full_cov`.
+    '''
+    with zs.BayesianNet(observed) as model:
+        k_inv_scale = tf.get_variable('k_inv_scale', [hps.N_X], hps.dtype,
+                                      tf.zeros_initializer())
+        k_inv_scale = tf.exp(k_inv_scale)
+        Kzz = rbf(Z_pos, Z_pos, k_inv_scale)
+        Kxz = rbf(X_ph, Z_pos, k_inv_scale)
+        Kzz_chol = tf.cholesky(Kzz)
+        fZ = zs.MultivariateNormalCholesky(
+            'fZ', tf.zeros([hps.N_Z], dtype=hps.dtype), Kzz_chol,
+            n_samples=n_particles)
+
+        # Model Y|Z.
+        # Mean[Y|Z] = Kxz @ inv(Kzz) @ fZ
+        # Cov[Y|Z] = Kxx - Kxz @ inv(Kzz) @ Kzx + noise_level * I
+        noise_level = tf.get_variable(
+            'noise_level', dtype=hps.dtype,
+            initializer=tf.constant(0.5, dtype=hps.dtype))
+        noise_level = tf.nn.softplus(noise_level)
+
+        # With ill-conditioned Kzz, the inverse is often asymmetric, which
+        # breaks further cholesky decomposition. We compute a symmetric one.
+        Kzz_chol_inv = tf.matrix_triangular_solve(
+            Kzz_chol, tf.eye(hps.N_Z, dtype=hps.dtype))
+        Kzz_inv = tf.matmul(tf.transpose(Kzz_chol_inv), Kzz_chol_inv)
+
+        Kxziz = tf.matmul(Kxz, Kzz_inv)
+        mean_y = tf.matmul(fZ, tf.transpose(Kxziz))  # [N_particles, N_X]
+
+        cov_xz = rbf(X_ph, X_ph, k_inv_scale) - \
+            tf.matmul(Kxziz, tf.transpose(Kxz))
+        cov_y = cov_xz + noise_level * tf.eye(
+            tf.shape(X_ph)[0], dtype=hps.dtype)
+        if full_cov:
+            cov_y_chol = tf.tile(tf.expand_dims(tf.cholesky(cov_y), 0),
+                                 [n_particles, 1, 1])
+            Y = zs.MultivariateNormalCholesky('Y', mean_y, cov_y_chol)
+        else:
+            std_y = tf.sqrt(tf.matrix_diag_part(cov_y))
+            std_y = tf.tile(tf.expand_dims(std_y, 0), [n_particles, 1])
+            Y = zs.Normal('Y', mean=mean_y, std=std_y, group_ndims=1)
+
+    return model, Y
+
+
+def build_variational(hps, n_particles):
+    with zs.BayesianNet() as variational:
+        Z_mean = tf.get_variable(
+            'Z/mean', [hps.N_Z], hps.dtype, tf.zeros_initializer())
+        Z_cov_raw = tf.get_variable(
+            'Z/cov_raw', initializer=tf.eye(hps.N_Z, dtype=hps.dtype))
+        Z_cov_tril = tf.matrix_set_diag(
+            tf.matrix_band_part(Z_cov_raw, -1, 0),
+            tf.nn.softplus(tf.matrix_diag_part(Z_cov_raw)))
```

Comment:
Why set diag again? To ensure it's positive? I don't get it.

Reply:
Yes. [By definition](https://en.wikipedia.org/wiki/Cholesky_decomposition) the diagonal part of the Cholesky decomposition should be positive, and we asserted this in `MultivariateNormalCholesky` when computing pdf, so we can write `sum(log(diag_part))` instead of `log(prod(diag_part))`.

Reply:
I see.